Add chainer.config.dtype and use it in initializers and dataset loaders.

beam2d opened the pull request:
This PR aims at making it easy to use fp16. By setting an environment variable `CHAINER_DTYPE=float16`, you can let weight initializers and datasets generate fp16 arrays by default instead of fp32. It makes each model definition independent from which dtype to use.

I confirmed that MNIST example runs without modifying other parts (except the eps parameter of Adam, whose default value is too small for fp16). This change does not alter the behavior of existing code unless `CHAINER_DTYPE` is set.

kmaehashi commented on chainer/__init__.py:
```diff
@@ -150,6 +150,7 @@ def is_arrays_compatible(arrays):
 global_config.use_ideep = os.environ.get('CHAINER_USE_IDEEP', 'never')
 global_config.lazy_grad_sum = bool(int(
     os.environ.get('CHAINER_LAZY_GRAD_SUM', '0')))
+global_config.dtype = numpy.dtype(os.environ.get('CHAINER_DTYPE', 'float32'))
```
It's better to show a nice error message if user specified wrong value.

beam2d replied:
Added a check for the value of CHAINER_DTYPE.

kmaehashi commented on chainer/datasets/cifar.py:
```diff
@@ -130,20 +134,21 @@ def load(archive, file_name):
     else:
         raise ValueError('invalid ndim for CIFAR dataset')
-    images = images.astype(numpy.float32)
+    dtype = chainer._get_dtype(dtype)
```
Need `import chainer` and `_get_dtype` -> `get_dtype`

beam2d replied:
fixed

kmaehashi commented:
Besides these fixes, we need to fix code that uses fp32 by default. e.g., https://github.com/chainer/chainer/blob/v4.0.0rc1/chainer/links/normalization/batch_normalization.py#L65

beam2d commented:
I will check other parts of Chainer (including BN you wrote) that should be fixed.

beam2d commented:
I checked functions and links that should be fixed to support fp16 mode, and found that there are so many 
things to fix. I feel it's better to fix them in separate PRs to make review quicker. I will make an issue to manage them.

kmaehashi commented:
@beam2d How do you want to proceed with this PR? I think it's OK to merge this first, then fix other functions and links (#4582).

beam2d commented:
I think it's good to first merge this PR and then proceed to #4582.

kmaehashi commented:
Jenkins, test this please.

beam2d commented:
I found that adding `.. currentmodule:: chainer` resolves the problem and still keeps `initializers` appear at the module index. I fixed it and rebased to the current master.

kmaehashi commented:
LGTM!
